robotbase/src/main/java/com/team2357/frc2022/subsystems/TurretSubsystem.java

```diff
@@ -0,0 +1,241 @@
+        m_pidController.setSmartMotionAllowedClosedLoopError(m_config.m_turretMotorAllowedError, smartMotionSlot);
+    }
+
+    public void setTurretSpeed(double speed) {
+        m_turretMotor.set(speed);
+    }
+
+    public void resetHeading() {
+        m_turretMotor.getEncoder().setPosition(0);
+    }
+
+    public double getTurretRotations() {
+        return m_turretMotor.getEncoder().getPosition();
+    }
+
+    public void setTurretPosition(double position) {
+        m_setPoint = position;
+        m_pidController.setReference(m_setPoint, CANSparkMax.ControlType.kSmartMotion);
+    }
+
+    public boolean atSetPoint() {
+        return com.team2357.frc2022.util.Utility.isWithinTolerance(getTurretRotations(), m_setPoint,
+                m_config.m_turretMotorAllowedError);
+    }
+
+    public boolean atSetPoint(double setPoint) {
+        return com.team2357.frc2022.util.Utility.isWithinTolerance(getTurretRotations(), setPoint,
+                m_config.m_turretMotorAllowedError);
+    }
+
+    public boolean isOnZero() {
+        boolean isMagnetDetected = false;
+
+        if (m_arduinoHallEffectSensor.isConnected()) {
+            isMagnetDetected = !m_arduinoHallEffectSensor
+                    .getDeviceFieldBoolean(Constants.ARDUINO.TURRET_HALL_SENSOR_NAME, "state");
+        }
+
+        if (isMagnetDetected) {
+            m_isZeroed = true;
+        }
+
+        return isMagnetDetected;
+    }
+
+    @Override
+    public void periodic() {
+        if (isClosedLoopEnabled()) {
+            closedLoopPeriodic();
+        }
+
+        // For tuning
+        SmartDashboard.putNumber("SetPoint", m_setPoint);
+        SmartDashboard.putNumber("Encoder Pos", getTurretRotations());
+        SmartDashboard.putNumber("Encoder Vel", m_turretMotor.getEncoder().getVelocity());
+        SmartDashboard.putNumber("Output", m_turretMotor.getAppliedOutput());
+    }
+
+    public boolean hasTarget() {
+        return m_currentTarget != null;
+    }
+
+    public boolean isTargetLocked() {
+        if (!isClosedLoopEnabled()) {
+            return false;
+        }
+        return atSetPoint();
+    }
+
+    @Override
+    public void setClosedLoopEnabled(boolean enabled) {
+        if (enabled) {
+            return;
+        }
+        disableClosedLoop();
+    }
+
+    public void enableClosedLoop(VisionTargetSupplier targetSupplier) {
+        m_targetSupplier = targetSupplier;
+        super.setClosedLoopEnabled(true);
+    }
+
+    public void disableClosedLoop() {
+        super.setClosedLoopEnabled(false);
+        m_targetSupplier = null;
+        m_turretMotor.set(0);
+    }
+
+    public void closedLoopPeriodic() {
+        m_currentTarget = m_targetSupplier.getAsVisionTarget();
+
+        if (m_isZeroed) {
+            if (m_currentTarget != null) {
+
+                double setPoint = calculateSetPoint(m_currentTarget);
+                setTurretPosition(setPoint);
+            } else {
+                lookForTarget();
```

Comment:
Let's not look for a target, let's just return an error state saying we couldn't find a target and make the operator try again when they've rotated the turret for the target to be in view.